File: dotnet/src/Agents/Core/History/ChatHistoryTruncationReducer.cs
// Copyright (c) Microsoft. All rights reserved.
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.SemanticKernel.ChatCompletion;

namespace Microsoft.SemanticKernel.Agents.History;

/// <summary>
/// Truncate the chat history to the target message count.
/// </summary>
/// <remarks>
/// Truncation will always Will avoid orphaning function-content as  the presence of
/// a function-call _must_ be followed by a function-result.  When a threshold count is
/// is provided (recommended), truncation will scan within the threshold window in an attempt to
/// avoid orphaning a user message from an assistant response.
/// </remarks>
internal class ChatHistoryTruncationReducer : IChatHistoryReducer
{
    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(nameof(ChatHistoryTruncationReducer), this._thresholdCount, this._targetCount);

    /// <inheritdoc/>
    public Task<IEnumerable<ChatMessageContent>?> ReduceAsync(IReadOnlyList<ChatMessageContent> history, CancellationToken cancellationToken = default)
    {
        // First pass to determine the truncation index
        int truncationIndex = LocateReductionIndex(history, this._targetCount, this._thresholdCount);

        IEnumerable<ChatMessageContent>? truncatedHistory = null;

        if (truncationIndex > 0)
        {
            // Second pass to truncate the history
            truncatedHistory = history.Extract(truncationIndex);
        }

        return Task.FromResult(truncatedHistory);
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ChatHistoryTruncationReducer"/> class.
    /// </summary>
    /// <param name="targetCount">The desired number of target messages after reduction.</param>
    /// <param name="thresholdCount">An optional number of messages beyond the 'targetCount' that must be present in order to trigger reduction/</param>
    /// <remarks>
    /// While the 'thresholdCount' is optional, it is recommended to provided so that reduction is not triggered
    /// for every incremental addition to the chat history beyond the 'targetCount'.
    /// </remarks>>
    public ChatHistoryTruncationReducer(int targetCount, int? thresholdCount = null)
    {
        Verify.True(targetCount > 0, "Target message count must be greater than zero.");
        Verify.True(!thresholdCount.HasValue || thresholdCount > 0, "The reduction threshold length must be greater than zero.");

        this._targetCount = targetCount;

        this._thresholdCount = thresholdCount ?? 0;
    }

    private static int LocateReductionIndex(IReadOnlyList<ChatMessageContent> history, int targetCount, int thresholdCount)
    {
        // Compute the index of the truncation threshold
        int thresholdIndex = history.Count - thresholdCount - targetCount;

        if (thresholdIndex <= 0)
        {
            // History is too short to truncate
            return 0;
        }

        // Compute the index of truncation target
        int messageIndex = history.Count - targetCount;

        // Skip function related content
        while (messageIndex >= 0)
        {
            if (!history[messageIndex].Items.Any(i => i is FunctionCallContent || i is FunctionResultContent))
            {
                break;
            }

            --messageIndex;
        }

        // Capture the earliest non-function related message
        int targetIndex = messageIndex;

        // Scan for user message within truncation range to maximize chat cohesion
        while (messageIndex >= thresholdIndex)
        {
            // A user message provides a superb truncation point
            if (history[messageIndex].Role == AuthorRole.User)
            {
                return messageIndex;
            }

            --messageIndex;
        }

        // No user message found, fallback to the earliest non-function related message
        return targetIndex;
    }

    private readonly int _thresholdCount;
    private readonly int _targetCount;
}

File: python/semantic_kernel/agents/history/chat_history_truncation_reducer.py
# Copyright (c) Microsoft. All rights reserved.

import logging
from typing import Any

from pydantic import Field

from semantic_kernel.agents.history.chat_history_reducer import ChatHistoryReducer
from semantic_kernel.agents.history.chat_history_reducer_extensions import extract_range, locate_safe_reduction_index
from semantic_kernel.contents.chat_message_content import ChatMessageContent
from semantic_kernel.kernel_pydantic import KernelBaseModel

logger = logging.getLogger(__name__)


class ChatHistoryTruncationReducer(KernelBaseModel, ChatHistoryReducer):
    """Truncate the chat history to the target message count.

    Avoids orphaning user messages or function-calls/results within a threshold window.
    """

    target_count: int = Field(..., gt=0, description="The target message count to reduce the chat history to.")
    threshold_count: int = Field(default=0, ge=0, description="The threshold count to avoid orphaning messages.")

    def __init__(self, target_count: int, threshold_count: int | None = None):
        """Initialize the truncation reducer."""
        args: dict[str, Any] = {
            "target_count": target_count,
        }

        if threshold_count is not None:
            args["threshold_count"] = threshold_count

        super().__init__(**args)

    def __eq__(self, other: object) -> bool:
        """Return whether this instance is equal to another."""
        if not isinstance(other, ChatHistoryTruncationReducer):
            return False
        return self.threshold_count == other.threshold_count and self.target_count == other.target_count

    def __hash__(self) -> int:
        """Return a hash code for this instance."""
        return hash((self.__class__.__name__, self.threshold_count, self.target_count))

    async def reduce(self, history: list[ChatMessageContent]) -> list[ChatMessageContent] | None:
        """Reduce the chat history to the target message count."""
        if len(history) <= self.target_count + (self.threshold_count or 0):
            # No need to reduce
            return None

        logger.info("Performing chat history truncation check...")

        truncation_index = locate_safe_reduction_index(history, self.target_count, self.threshold_count)
        if truncation_index < 0:
            logger.info(
                f"No truncation index found. Target count: {self.target_count}, Threshold count: {self.threshold_count}"
            )
            return None

        logger.info(f"Valid truncation index found. Truncating history to {truncation_index} messages.")

        return extract_range(history, start=truncation_index)
